### PR TITLE
Register UI interfaces and remove non-essential $run() arguments

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #562

Includes all of #562 .  Made a new PR to add some final code changes that I did not explain well.

PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

### CODE REVIEW

File: R/plumber.R [hunk 1]
```diff
@@ -902,6 +876,54 @@ plumber <- R6Class(
     setErrorHandler = function(fun){
       private$errorHandler <- fun
     },
+    #' @description Add a function to customize what is returned in `$apiSpec()`.
+    #' @param api_fun This function should accept an OpenAPI Specification autogenerated by `plumber`.  It should return a list object in valid OpenAPI Specification format. This will not be validated.
```
Comment (scottmmjackson) [inline R/plumber.R:880]: The docs here feel like they could be fleshed out a little bit.
- What's the exact signature we expect?
- Is there an example input and output?

What it sounds like, and this feels a bit wrong, is that it receives an OpenAPI JSON string and expects a listified version of it back, which sounds a lot like `fromJSON`, with all the weird array encapsulation things implied.

File: R/plumber.R [hunk 2]
```diff
@@ -902,6 +876,54 @@ plumber <- R6Class(
...
+      if (is.null(api_fun)) {
+        api_fun <- identity
+      }
+      if (!is.function(api_fun)) {
+        stop("`api_fun` must be a function")
+      }
+      private$apiHandler <- api_fun
+    },
+    #' @description Set UI to use for API
+    #' @param ui a character value or a logical value. Default to `plumber.ui` option value.
```
Comment (scottmmjackson) [inline R/plumber.R:891]: There's not a list of valid values in here, nor a hint as to where to find them.

File: R/plumber.R [hunk 3]
```diff
@@ -902,6 +876,54 @@ plumber <- R6Class(
...
+    #' @param callback a callback function for taking action on UI url.
```
Comment (scottmmjackson) [inline R/plumber.R:892]: Is this a Rook function? We should say that if so.

File: tests/testthat/test-options.R [hunk 4]
```diff
@@ -34,6 +34,8 @@ test_that("all options used are `options_plumber()` parameters", {
   }
   options_used <- unique(sort(gsub("getOption|\\(|\"|,|'|\\)", "", matches)))
   plumber_options_used <- grep("^plumber", options_used, value = TRUE)
+  deprecated_options <-  c("plumber.swagger.url")
+  plumber_options_used <- plumber_options_used[!plumber_options_used %in% deprecated_options]
```
Comment (scottmmjackson) [inline tests/testthat/test-options.R:38]: Does order of operations prefer `%in%` to `!`? It feels a bit like you might be coercing the string to a boolean, negating it, and checking for the boolean's presence in `deprecated_options` which would produce a false positive- maybe? I could be wrong here.

Comment (schloerke) [reply tests/testthat/test-options.R:38]: `%in%` has precedence. Will add parens for clarity

File: R/plumber.R (hunk 3 above)
Comment (schloerke) [reply R/plumber.R:892]: No. Needs more description

File: R/plumber.R (hunk 1 above)
Comment (schloerke) [reply R/plumber.R:880]: > with all the weird array encapsulation things implied

Yup

File: R/plumber.R [hunk 5]
```diff
@@ -258,52 +264,63 @@ plumber <- R6Class(
...
+      # Checks must be kept for >= 2 yrs after plumber v1.0.0 release date
+      if (!missing(debug)) {
+        message("`$run(debug)` has been deprecated in v1.0.0 and will be removed in a coming release. Please use `$setDebug(debug)`")
+        self$setDebug(debug)
+      }
+      if (!missing(swagger)) {
+        if (is.function(swagger)) {
+          # between v1.0.0 and v0.4.6
+          message("`$run(swagger)` has been deprecated in v1.0.0 and will be removed in a coming release. To alter the swagger spec, please use `$apiSpecTransform(api_fun)`")
+          self$apiSpecTransform(swagger)
```
Comment (schloerke) [inline R/plumber.R:294]: Should `$apiSpecTransform()` be renamed to `$setApiSpecTransform()`?  Same argument within the tidy API when updating the default serializer

File: R/plumber.R [hunk 6]
```diff
@@ -258,52 +264,63 @@ plumber <- R6Class(
...
+            message("`$run(swagger)` has been deprecated in v1.0.0 and will be removed in a coming release. Please use `$setUI(ui)`")
+            self$setUI(swagger)
+          } else {
+            # $setUI has been called (other than during initialization).
+            # Believe that it is the correct behavior
+            # Warn about updating the run method
+            message(
+              "`$run(swagger)` has been deprecated in v1.0.0 and will be removed in a coming release.\n",
+              "The plumber UI has already been set. Ignoring `swagger` parameter.\n",
+              "Please use update your `$run()` method."
```
Comment (schloerke) [inline R/plumber.R:307]:
```suggestion
              "Please update your `$run()` method."
```

File: R/plumber.R (hunk 5 above)
Comment (meztez) [reply R/plumber.R:294]: https://github.com/rstudio/plumber/pull/562#discussion_r453889135

Comment (cpsievert) [reply R/plumber.R:294]: After discussing larger API naming decisions, we're now thinking `$apiSpec()` should  be `$getApiSpec()` and `$apiSpecTransform()` should be `$setApiSpec()` (and should allow either a function or spec R list object)

Comment (cpsievert) [reply R/plumber.R:294]: Then, after #590 is merged, we'd add `pr_set_api_spec()` (it doesn't seem worth adding the get method)

File: R/plumber.R [hunk 7]
```diff
@@ -258,52 +264,63 @@ plumber <- R6Class(
...
       if (isTRUE(private$disable_run)) {
         stop("Plumber router `$run()` method should not be called while `plumb()`ing a file")
       }
 
-      port <- findPort(port)
+      # Legacy support for RStudio pro products.
+      # Checks must be kept for >= 2 yrs after plumber v1.0.0 release date
+      if (!missing(debug)) {
+        message("`$run(debug)` has been deprecated in v1.0.0 and will be removed in a coming release. Please use `$setDebug(debug)`")
+        self$setDebug(debug)
```
Comment (cpsievert) [inline R/plumber.R:288]: Make a `pr_set_debug()` after #590

File: R/plumber.R [hunk 8]
```diff
@@ -258,52 +264,63 @@ plumber <- R6Class(
...
-    #' }
     run = function(
       host = '127.0.0.1',
       port = getOption('plumber.port'),
-      swagger = interactive(),
-      debug = interactive(),
-      swaggerCallback = getOption('plumber.swagger.url', NULL)
+      debug = stop("deprecated"),
+      swagger = stop("deprecated"),
+      swaggerCallback = stop("deprecated")
```
Comment (cpsievert) [inline R/plumber.R:277]:
```suggestion
      swagger = stop("deprecated"),
      debug = stop("deprecated"),
      swaggerCallback = stop("deprecated")
```

File: R/plumber.R (hunk 5 above)
Comment (cpsievert) [inline R/plumber.R:294]:
```suggestion
          self$apiSpecTransform(swagger)
          self$setUI(TRUE)
```

File: R/ui.R [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+#'
+#' `ui` list expects the following values
+#' \describe{
+#' \item{package}{Name of the package required for the UI.}
+#' \item{name}{Name of the UI.}
+#' \item{index}{A function that returns the HTML content of the landing page of the UI.}
+#' \item{static}{A function that returns the path to the assets (images, javascript, css, fonts) the UI will use.}
+#' }
+#' @export
+add_ui <- function(ui) {
```
Comment (cpsievert) [inline R/ui.R:142]: Maybe a name like `register_api_ui()` would be better?

File: R/plumber.R [hunk 10]
```diff
@@ -902,6 +876,54 @@ plumber <- R6Class(
...
+      }
+      private$apiHandler <- api_fun
+    },
+    #' @description Set UI to use for API
+    #' @param ui a character value or a logical value. Default to `plumber.ui` option value.
+    #' @param callback a callback function for taking action on UI url.
+    #' @param ... Other params to be passed to ui functions.
+    setUI = function(
+      ui = getOption("plumber.ui", TRUE),
+      callback = getOption('plumber.ui.callback', getOption('plumber.swagger.url', NULL)),
```
Comment (cpsievert) [inline R/plumber.R:896]: Consider providing a `setUICallback` method

**Comment by blairj09** [comment]
Currently, an API built against this PR published to RStudio Connect results in the following:
```
2020/07/21 20:31:22.911759238 Using Packrat dir /opt/rstudio-connect/mnt/app/packrat/lib/x86_64-pc-linux-gnu/3.6.0
2020/07/21 20:31:23.477196448 `$run(swagger)` has been deprecated in v1.0.0 and will be removed in a coming release. Please use `$setUI(ui)`
2020/07/21 20:31:23.477397750 Running plumber API at http://127.0.0.1:38837
2020/07/21 20:31:23.484268239 Error in value[[3L]](cond) : 
2020/07/21 20:31:23.484278045   yaml must be installed for the yaml serializer to work
2020/07/21 20:31:23.484308581 Calls: local ... tryCatch -> tryCatchList -> tryCatchOne -> <Anonymous>
2020/07/21 20:31:23.484309648 Execution halted
```

It looks like we'll need to add `yaml` as an explicit dependency.

Once `library(yaml)` was added to the API (in order to get RStudio Connect to acknowledge it) the UI interface rendered as expected 🙌 
image

**Comment by schloerke** [comment]
@blairj09 The `yaml` error is new behavior from parsers and should disappear when https://github.com/rstudio/plumber/pull/584 is merged. (Parsers will now be "opt-in", rather than "parse any/everything")

File: R/plumber.R (hunk 5 above)
Comment (schloerke) [reply R/plumber.R:294]: Opting against this as it is enabled by default and they may have set a UI sometime earlier.  Keep the message about `swagger` can not be a function.

File: R/ui.R [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  .globals$UIs[[ui$name]]$mount <- mount_ui_func
+  .globals$UIs[[ui$name]]$unmount <- unmount_ui_func
+
+  invisible()
+}
+#' @export
+#' @rdname register_ui
+registered_ui <- function() {
```
Comment (cpsievert) [inline R/ui.R:190]: Plural

File: R/ui.R [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+  sort(names(.globals$UIs))
+}
+
+# TODO: Remove once UI load code moved to respective UI package
+swagger_ui <- list(
+  package = "swagger",
```
Comment (cpsievert) [inline R/ui.R:196]: Let's remove this package field (the `index` function should be enough to require/throw the package name)

File: R/ui.R [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+      # Override with arguments provided live with URI (i.e. index.html?version=2)
+      args <- utils::modifyList(args_index, list(...))
+      # Remove default arguments req and res
+      args <- args[!(names(args) %in% c("req", "res"))]
+      do.call(ui$index, args)
+    }
+    for (path in ui_path) {
+      pr$handle("GET", path, ui_index,  serializer = serializer_html())
+    }
+    pr$mount(ui_root, PlumberStatic$new(ui$static(...)))
```
Comment (cpsievert) [inline R/ui.R:172]: Do this via a full subrouter (not just mounting to the original)

Comment (cpsievert) [reply R/ui.R:172]: Also, we should emit a message/warning if there is already something mounted here

File: R/plumber.R [hunk 14]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+    #'
+    #'  The value returned will not be validated for OAS compatibility.
+    set_api_spec = function(api = NULL) {
+      api_fun <-
+        if (is.null(api)) {
+          identity
+        } else {
+          if (!is.function(api)) {
+            # function to return the api object
+            function(...) {
```
Comment (cpsievert) [inline R/plumber.R:987]: Let's be a bit clearer this should be single arg

```suggestion
            function(x) {
```

File: R/pr_set.R [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+#'
+#' pr() %>%
+#'   pr_get("/hi", function() "Hello") %>%
+#'   pr_set_404(handler_404) %>%
+#'   pr_run()
+#' }
+#'
+#' @export
+pr_set_404 <- function(pr,
+                   fun) {
```
Comment (cpsievert) [inline R/pr_set.R:44]:
```suggestion
pr_set_404 <- function(pr, fun) {
```

File: R/pr_set.R [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+#' default serializer to the function supplied via \code{serializer}
+#'
+#' @template param_pr
+#' @param serializer A serializer function
+#'
+#' @return The Plumber router with the new default serializer
+#'
+#' @export
+pr_set_serializer <- function(pr,
+                          serializer) {
```
Comment (cpsievert) [inline R/pr_set.R:13]:
```suggestion
pr_set_serializer <- function(pr, serializer) {
```

File: R/pr_set.R [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+#' }
+#'
+#' pr() %>%
+#'   pr_get("/error", function() log("a")) %>%
+#'   pr_set_error(handler_error) %>%
+#'   pr_run()
+#' }
+#' @export
+pr_set_error <- function(pr,
+                     fun) {
```
Comment (cpsievert) [inline R/pr_set.R:72]:
```suggestion
pr_set_error <- function(pr, fun) {
```

File: R/pr_set.R [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+#' file will be served on paths `/openapi.json` and `/swagger.json`. Swagger UI
+#' will be served on paths `/__swagger__/index.html` and `/__swagger__/`. When
+#' using a function, it will receive the Plumber router as the first parameter
+#' and current OpenAPI Specification as the second. This function should return a
+#' list containing OpenAPI Specification.
+#' See \url{http://spec.openapis.org/oas/v3.0.3}
+#'
+#' @template param_pr
+
+#' @param ui a character value or a logical value. Default to `plumber.ui` option value.
```
Comment (cpsievert) [inline R/pr_set.R:125]: Lets make sure to link to `options_plumber` in these scenarios

Comment (cpsievert) [reply R/pr_set.R:125]: And/or add comment to see the example section for examples of different values

File: R/pr_set.R [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+#'
+#' @template param_pr
+#' @template pr_set_api_spec__api
+#' @return The Plumber router with the new OpenAPI Specification object or function.
+#' @export
+#' @examples
+#' \dontrun{
+#' # Set the API Spec to a function to use the auto-generated OAS object
+#' pr() %>%
+#'   pr_set_ui(TRUE) %>%
```
Comment (cpsievert) [inline R/pr_set.R:214]:
```suggestion
```

File: R/pr_set.R [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+#'   pr_set_api_spec(function(spec) {
+#'     spec$info$title <- Sys.time()
+#'     spec
+#'   }) %>%
+#'   pr_get("/plus/<a:int>/<b:int>", function(a, b) { a + b }) %>%
+#'   pr_run()
+#'
+#' # Set the API Spec using an object
+#' pr() %>%
+#'   pr_set_ui(TRUE) %>%
```
Comment (cpsievert) [inline R/pr_set.R:224]:
```suggestion
```

File: R/pr_set.R [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+#' Plumber router [pr_run()] method is executed using default `plumber.ui.callback` option.
+#'
+#' @template param_pr
+#' @param callback a callback function for taking action on UI url.
+#' @return The Plumber router with the new UI callback setting.
+#' @export
+#' @examples
+#' \dontrun{
+#' pr() %>%
+#'   pr_set_ui(TRUE) %>%
```
Comment (cpsievert) [inline R/pr_set.R:185]:
```suggestion
```

File: R/plumber.R [hunk 22]
```diff
@@ -248,58 +257,72 @@ plumber <- R6Class(
...
-    #' swagger <- function(pr_, spec) {
-    #'   spec$servers[[1]]$description <- "MyCustomAPISpec"
-    #'   spec
-    #' }
-    #' pr$run(swagger = swagger)
-    #' }
+    #' This value does not need to be explicitly assigned.
+    #' @param debug Deprecated. See `$set_debug()`
+    #' @param swagger Deprecated. See `$set_ui(ui)` or `$set_api_spec()`
+    #' @param swaggerCallback Deprecated. See `$set_ui(callback)`
```
Comment (cpsievert) [inline R/plumber.R:272]:
```suggestion
    #' @param swaggerCallback Deprecated. See `$set_ui_callback()`
```

File: NEWS.md [hunk 23]
```diff
@@ -18,27 +18,32 @@ plumber 1.0.0
...
-
 ### Breaking changes
 
-* Plumber's swagger definition is now defined using
-  [OpenAPI 3](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md),
+* Plumber's OpenAPI Specification is now defined using
+  [OpenAPI 3](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md),
   upgrading from Swagger Specification. (#365)
 
+* Plumber router `$run()` method arguments `swagger`, `swaggerCallback` and `...` deprecated. User interface and url callback are now enabled by default and managed through Plumber router `$set_ui()` method and option `plumber.ui`. (@meztez, #562)
```
Comment (cpsievert) [inline NEWS.md:27]:
```suggestion
* Plumber router `$run()` method arguments `swagger`, `swaggerCallback` and `debug` are now deprecated. User interface and url callback are now enabled by default and managed through Plumber router `$set_ui()`, `$set_ui_callback()`, and `$set_debug()` methods and options `plumber.ui` and `plumber.ui.callback`. (@meztez, #562)
```

File: R/plumber.R [hunk 24]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+      if (isTRUE(ui)) {
+        ui <- "swagger"
+      }
+      if (is.character(ui) && is_ui_available(ui)) {
+        enabled <- TRUE
+      } else {
+        enabled <- FALSE
+        ui <- "__not_enabled__"
+      }
+      private$ui_info = list(
```
Comment (cpsievert) [inline R/plumber.R:929]:
```suggestion
      private$ui_info <- list(
```

File: R/plumber.R [hunk 25]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+    #' Plumber router [pr_run()] method is executed using default `plumber.ui.callback` option.
+    #'
+    #' See also: [pr_set_ui_callback()]
+    #' @param callback a callback function for taking action on UI url. (Also accepts `NULL` values to disable the `callback`.)
+    set_ui_callback = function(
+      callback = getOption('plumber.ui.callback', getOption('plumber.swagger.url', NULL))
+    ) {
+      # Use callback when defined
+      if (!length(callback) || !is.function(callback)) {
+        callback = as.null
```
Comment (cpsievert) [inline R/plumber.R:948]:
```suggestion
        callback <- function(...) { NULL }
```

File: R/plumber.R [hunk 26]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+      }
+      private$ui_callback = callback
```
Comment (cpsievert) [inline R/plumber.R:950]:
```suggestion
      private$ui_callback <- callback
```

File: R/plumber.R [hunk 27]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+    #'
+    #' See also: [pr_set_ui_callback()]
+    #' @param callback a callback function for taking action on UI url. (Also accepts `NULL` values to disable the `callback`.)
+    set_ui_callback = function(
+      callback = getOption('plumber.ui.callback', getOption('plumber.swagger.url', NULL))
+    ) {
+      # Use callback when defined
+      if (!length(callback) || !is.function(callback)) {
+        callback = as.null
+      }
```
Comment (cpsievert) [inline R/plumber.R:949]: Add a check for one or more args on `callback`

File: R/plumber.R [hunk 28]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+      private$ui_callback = callback
+    },
+    #' @description Set debug value to include error messages
+    #' @param debug `TRUE` provides more insight into your API errors.
+    #' @seealso [pr_set_debug()]
+    set_debug = function(debug = interactive()) {
+      stopifnot(length(debug) == 1)
+      private$debug = isTRUE(debug)
```
Comment (cpsievert) [inline R/plumber.R:957]:
```suggestion
      private$debug <- isTRUE(debug)
```

File: R/plumber.R [hunk 29]
```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
...
+    },
     #' @description Add a filter to plumber router
     #' @param name a character string. Name of filter
     #' @param expr an expr that resolve to a filter function or a filter function
     #' @param serializer a serializer function
+    #' @seealso [pr_filter()]
```
Comment (cpsievert) [inline R/plumber.R:963]: Move to description

**Comment by schloerke** [comment]
- [ ] Add `pr_set_parsers()` method
- [ ] `$setParsers()` -> `set_parsers()` ?

**Comment by schloerke** [comment]
Squash Merging this PR on a Friday afternoon. Need to get it into people hands for them to make noise. 😉

**Pull request merged**